File: src/main/java/com/yuukaze/i18next/ui/tabs/TableView.java
package com.yuukaze.i18next.ui.tabs;

import com.intellij.openapi.project.Project;
import com.intellij.ui.JBColor;
import com.intellij.ui.border.CustomLineBorder;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.ui.table.JBTable;
import com.yuukaze.i18next.model.*;
import com.yuukaze.i18next.model.table.TableModelTranslator;
import com.yuukaze.i18next.service.DataStore;
import com.yuukaze.i18next.ui.components.RootKeyTree;
import com.yuukaze.i18next.ui.dialog.EditDialog;
import com.yuukaze.i18next.ui.listener.DeleteKeyListener;
import com.yuukaze.i18next.ui.listener.PopupClickListener;
import com.yuukaze.i18next.ui.renderer.CustomTableHeaderCellRenderer;
import com.yuukaze.i18next.ui.renderer.CustomTableCellRenderer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.event.MouseEvent;
import java.util.ResourceBundle;

/**
 * Shows translation state as table.
 *
 * @author marhali
 */
public class TableView implements DataSynchronizer {

    private final Project project;

    private JPanel rootPanel;
    private JPanel containerPanel;
    private JScrollPane keyTreePane;
    private final RootKeyTree rootKeyTree;
    private final JBTable table;

    public TableView(Project project) {
        this.project = project;

        table = new JBTable();
        table.getEmptyText().setText(ResourceBundle.getBundle("messages").getString("view.empty"));
        table.addMouseListener(new PopupClickListener(this::handlePopup));
        table.addKeyListener(new DeleteKeyListener(handleDeleteKey()));
        table.setDefaultRenderer(String.class, new CustomTableCellRenderer());
        table.getTableHeader().setReorderingAllowed(false);
        table.getTableHeader().setDefaultRenderer(new CustomTableHeaderCellRenderer());

        JBScrollPane scrollPane = new JBScrollPane(table);
        scrollPane.setBorder(new CustomLineBorder(JBColor.border(), 0, 1, 0, 0));
        containerPanel.add(scrollPane);

        rootKeyTree = new RootKeyTree(project);
        keyTreePane.setViewportView(rootKeyTree);
    }

    private void handlePopup(MouseEvent e) {
        int row = table.rowAtPoint(e.getPoint());

        if (row >= 0) {
            String fullPath = String.valueOf(table.getValueAt(row, 0));
            LocalizedNode node = DataStore.getInstance(project).getTranslations().getNode(fullPath);

            if (node != null) {
                new EditDialog(project, new KeyedTranslation(fullPath, node.getValue())).showAndHandle();
            }
        }
    }

    private Runnable handleDeleteKey() {
        return () -> {
            for (int selectedRow : table.getSelectedRows()) {
                String fullPath = String.valueOf(table.getValueAt(selectedRow, 0));

                DataStore.getInstance(project).processUpdate(
                        new TranslationDelete(new KeyedTranslation(fullPath, null)));
            }
        };
    }

    @Override
    public void synchronize(@NotNull Translations translations, @Nullable String searchQuery) {
        table.setModel(new TableModelTranslator(translations, searchQuery, update ->
                DataStore.getInstance(project).processUpdate(update)));
    }

    public JPanel getRootPanel() {
        return rootPanel;
    }

    public JBTable getTable() {
        return table;
    }

    public RootKeyTree getRootKeyTree() {
        return rootKeyTree;
    }
}

File: src/main/kotlin/com/yuukaze/i18next/ui/tabs/TableView.kt
package com.yuukaze.i18next.ui.tabs

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.JBMenuItem
import com.intellij.openapi.ui.JBPopupMenu
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.ui.JBColor
import com.intellij.ui.border.CustomLineBorder
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.layout.panel
import com.intellij.ui.table.JBTable
import com.yuukaze.i18next.model.*
import com.yuukaze.i18next.model.table.TableModelTranslator
import com.yuukaze.i18next.service.DataStore
import com.yuukaze.i18next.service.getEasyI18nDataStore
import com.yuukaze.i18next.ui.components.RootKeyTree
import com.yuukaze.i18next.ui.dialog.MigrateDialog
import com.yuukaze.i18next.ui.listener.DeleteKeyListener
import com.yuukaze.i18next.ui.listener.DoubleClickListener
import com.yuukaze.i18next.ui.model.FilterUntranslatedModel
import com.yuukaze.i18next.ui.renderer.CustomTableCellRenderer
import com.yuukaze.i18next.ui.renderer.CustomTableHeaderCellRenderer
import java.awt.event.ActionEvent
import java.awt.event.MouseEvent
import java.beans.PropertyChangeListener
import java.util.*
import javax.swing.Action
import javax.swing.JPanel

/**
 * Shows translation state as table.
 */
class TableView(private val project: Project?) : DataSynchronizer {
  val rootPanel: JPanel
    get() = panel {
      row {
        scrollPane(rootKeyTree)
        scrollPane(table).constraints(pushX)
      }
    }
  val rootKeyTree: RootKeyTree
  val table: JBTable = JBTable()
  private val filterUntranslated = FilterUntranslatedModel()
  private val popupMenu = JBPopupMenu().let {
    it.add(JBMenuItem(TableViewAction("Edit...", this::handleEdit)))
    it.add(JBMenuItem(TableViewAction("Migrate...", this::handleMigrate)))
    it.add(JBMenuItem(TableViewAction("Delete", this::handleDelete)))
    it
  }

  init {
    table.emptyText.text =
      ResourceBundle.getBundle("messages").getString("view.empty")
    table.addKeyListener(DeleteKeyListener(handleDeleteKey()))
    table.setDefaultRenderer(
      String::class.java,
      CustomTableCellRenderer(project!!)
    )
    table.componentPopupMenu = popupMenu
    setupTableHeader()
    val scrollPane = JBScrollPane(table)
    scrollPane.border = CustomLineBorder(JBColor.border(), 0, 1, 0, 0)
    rootKeyTree = RootKeyTree(project)
  }

  private fun setupTableHeader() {
    val header = table.tableHeader
    header.reorderingAllowed = false
    header.defaultRenderer =
      CustomTableHeaderCellRenderer(header, filterUntranslated)
    header.addMouseListener(DoubleClickListener { e: MouseEvent ->
      handleFilter(
        e
      )
    })
  }

  private fun handleFilter(e: MouseEvent) {
    val col = table.columnAtPoint(e.point)
    val colName = table.getColumnName(col)
    if (col >= 0 && filterUntranslated.available.contains(colName)) {
      val current = filterUntranslated.toggle
      filterUntranslated.toggle = if (current == colName) null else colName
    }
  }

  private fun handleDeleteKey(): Runnable {
    return Runnable {
      for (selectedRow in table.selectedRows) {
        val fullPath = table.getValueAt(selectedRow, 0).toString()
        DataStore.getInstance(project).processUpdate(
          TranslationDelete(KeyedTranslation(fullPath, null))
        )
      }
    }
  }

  override fun synchronize(translations: Translations, searchQuery: String?) {
    table.model = TableModelTranslator(
      translations,
      searchQuery
    ) { update: TranslationUpdate? ->
      DataStore.getInstance(
        project
      ).processUpdate(update)
    }
  }

  private fun handleEdit(row: Int) {
    //TODO implement
  }

  private fun handleMigrate(row: Int) {
    MigrateDialog(project!!).showAndHandle()
  }

  private fun handleDelete(row: Int) {
    val key = table.getValueAt(row, 0).toString()
    JBPopupFactory.getInstance()
      .createConfirmation(
        "Delete key $key?", "Yes", "No",
        {
          val dataStore = project.getEasyI18nDataStore()
          dataStore.translations.nodes.removeChildren(key)
          dataStore.doSync()
        }, 0
      ).showCenteredInCurrentWindow(project!!)
  }

  inner class TableViewAction(
    private val name: String,
    val callback: (row: Int) -> Unit
  ) :
    Action {
    override fun actionPerformed(e: ActionEvent?) {
      callback(table.selectedRow)
    }

    override fun getValue(key: String?): Any? = when (key) {
      Action.NAME -> name
      Action.MNEMONIC_KEY -> key[0].toInt()
      else -> null
    }

    override fun putValue(key: String?, value: Any?) {}

    override fun setEnabled(b: Boolean) {}

    override fun isEnabled(): Boolean = true

    override fun addPropertyChangeListener(listener: PropertyChangeListener?) {}

    override fun removePropertyChangeListener(listener: PropertyChangeListener?) {}
  }
}